src/components/utils/src/timer.cc

```diff
@@ -44,175 +44,187 @@ CREATE_LOGGERPTR_GLOBAL(logger_, "Utils")
-
-  if (timer_settime(internal_timer, 0, &itimer, NULL) == kErrorCode) {
-    int error_code = errno;
-    UNUSED(error_code);
-    LOG4CXX_FATAL(logger_,
-                  "Can`t set timeout to posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return NULL;
-  }
-  return internal_timer;
-}
-
-bool StopPosixTimer(timer_t timer) {
-  LOG4CXX_AUTO_TRACE(logger_);
-  const int resultCode = timer_delete(timer);
-  if (kErrorCode == resultCode) {
-    int error_code = errno;
-    LOG4CXX_ERROR(logger_,
-                  "Can`t delete posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return false;
-  }
-  return true;
-}
-}  // namespace
-
-timer::Timer::Timer(const std::string& name, const TimerTask* task_for_tracking)
+timer::Timer::Timer(const std::string& name, TimerTask* task_for_tracking)
     : name_(name)
     , task_(task_for_tracking)
-    , repeatable_(false)
     , timeout_ms_(0u)
     , is_running_(false)
-    , timer_(NULL) {
+    , delegate_(this)
+    , thread_(threads::CreateThread(name_.c_str(), &delegate_)) {
   LOG4CXX_AUTO_TRACE(logger_);
   DCHECK(!name_.empty());
-  DCHECK(task_);
+  DCHECK(task_.get());
+  DCHECK(thread_);
 }
 
 timer::Timer::~Timer() {
   LOG4CXX_AUTO_TRACE(logger_);
   LOG4CXX_DEBUG(logger_, "Timer is to be destroyed " << name_);
-  Stop();
   sync_primitives::AutoLock auto_lock(task_lock_);
-  DCHECK(task_);
-  delete task_;
+
+  DCHECK(thread_);
+  thread_->join();
+  DeleteThread(thread_);
 }
 
 void timer::Timer::Start(const Milliseconds timeout, const bool repeatable) {
   LOG4CXX_AUTO_TRACE(logger_);
   sync_primitives::AutoLock auto_lock(lock_);
-  SetTimeoutUnsafe(timeout);
-  repeatable_ = repeatable;
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
+
+  LOG4CXX_DEBUG(logger_, "Prepare start timer for"
+                          << timeout
+                          << "ms, repetable = "
+                          << (repeatable ? "true" : "false"));
+
+  if (repeatable) {
+    delegate_.make_repeatable();
   }
-  StartUnsafe();
+
+  if (thread_->is_running()) {
+    LOG4CXX_INFO(logger_, "Restart timer in thread " << name_);
+    delegate_.ShouldBeRestarted();
+    UpdateTimeOut(timeout);
+  } else {
+    LOG4CXX_INFO(logger_, "Start new timer in thread " << name_);
+    UpdateTimeOut(timeout);
+    thread_->start();
+  }
+
+  delegate_.WaitUntilStart();
 }
 
 void timer::Timer::Stop() {
   LOG4CXX_AUTO_TRACE(logger_);
   sync_primitives::AutoLock auto_lock(lock_);
-  repeatable_ = false;
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK(stop_result);
+  DCHECK(thread_);
+  LOG4CXX_DEBUG(logger_, "Stopping timer  " << name_);
+  if (pthread_equal(pthread_self(), thread_->thread_handle())) {
+    // Thread can't stop itself , so it will suspend
+    Suspend();
+  } else {
+    thread_->join();
```

Comment:
@anosach-luxoft , to prevent possible situation when will be delete timer with working OnTimer.